Desktop v3

MadLittleMods opened the pull request:
- Update docs to add more detail on how to build
- Upgrade to nw.js v0.12.3
  - Fixes issue where the org overview pages had an infinite loading spinner: https://github.com/gitterHQ/desktop/issues/102
  - Fixes `Cmd+M` shortcut to minimize: https://github.com/gitterHQ/desktop/issues/91
  - Fixes tray icon missing on Ubuntu: https://github.com/gitterHQ/desktop/issues/101
- Fix focus issue on initial load with mac. Mac apps do not have focus when they start up
- Update `notifier.js` so that it no longer plays the native sound, fixes https://github.com/gitterHQ/desktop/issues/99
- Add avatars to notifications for Windows
- I assume this covers as the official release of the nw.js version of the mac app: https://github.com/gitterHQ/desktop/issues/47
- Generalize windows desktop build, see `./windows/README.md` and `./windows/build.js`. `./windows/build.bat` has also gotten some love but is deprecated and is only there until we are confident and settled with the node script.
- Remove startup option from Windows installer because it did not obey whatever you set it to(always enabled launch at startup): https://github.com/gitterHQ/desktop/issues/106
- Add context menu option to toggle launch on startup (tested on Mac and Windows): https://github.com/gitterHQ/desktop/issues/45
- Add "Show in Menu bar" toggle (this option existed previously but didn't do anything)
- Add `gulp check-path-safety-for-windows` to build flow so you don't get [unexpected blank screen when developing on Windows](https://github.com/gitterHQ/desktop/issues/59)
- Update `auto-update` params/logic so we can actually use cli params outside of the autoupdate process.
  - `--update-url`: Specify custom URL to check and download updates from
  - `--passthrough-remote-debugging-port`: So you can debug the new instances we pop throughout the update process
- Update auto-update logic to check on start and every 24-hours after that
- Add `packages -> osx` `package.json` manifest check entry to play more nicely with our platform checks.

---
### Notifications

Windows 10
![](http://i.imgur.com/3y2RlB3.png)

Windows 8.1
![](http://i.imgur.com/HObLa2w.png)

Mac OSX
![](https://i.imgur.com/BHLXShe.png)

Ubuntu 14.04.03
![](https://i.imgur.com/W0MwSq8.png)

---
### Debug nw.js from another PC.

Route `localhost:9222` to the external reachable local network. Here are instructions for setting it up on  Mac: https://github.com/nwjs/nw.js/issues/1444#issuecomment-179926199

`localhost:9222` -> `192.168.0.58:9222`

Then start up the nw.js app:

```
npm start -- --remote-debugging-port=9222
```

Visit `192.168.0.58:9222` on the other machine

These parameters also work on a final build.

**Windows**

```
"C:\Users\YOUR_USER\AppData\Local\Programs\Gitter\Gitter.exe" --remote-debugging-port=9222 --passthrough-remote-debugging-port=9222
```

**OSX**

```
open /Applications/Gitter.app --args --remote-debugging-port=9222 --passthrough-remote-debugging-port=9222
```

---
### Windows 10 notifications

Just to dump my knowledge here while it's fresh:

The [new notifications in Windows 10 are "toast" notifications](https://msdn.microsoft.com/en-us/library/windows/desktop/hh802762%28v=vs.85%29.aspx). They can be setup by creating a `.lnk`also know as "shortcut" in one of the Start menu folders such as "C:\Users\MLM\AppData\Roaming\Microsoft\Windows\Start Menu\Programs".

With `nw.js` you can use [`var gui = require('nw.gui'); gui.App.createShortcut(...);`_(read more here)_](https://github.com/nwjs/nw.js/wiki/Notification#toast-notification-requires-node-webkit--v011) but we decided to go with [`node-notifier`](https://www.npmjs.com/package/node-notifier) because it has a `sound: false` option also know as the "quiet" or "silent" option in their internals. Which can stop the native Windows 10 notification sound from chiming in.

Users can manually manage their notification preferences on a individual basis by opening "Notifications & action settings". http://i.imgur.com/DjNB2oL.png - Because we are using `node-notifier` the entry is called `toast` which is the vendor specific exe([find the source here](https://github.com/mikaelbr/node-notifier/issues/96)) they use to make the notifications. This does mean that the entry is shared across all apps that use `node-notifier` but the user shouldn't be touching it in the settings anyway because apps customize their own behaviour (hopefully conscientiously)

MadLittleMods committed:
Update to nw.js 0.12.3

MadLittleMods committed:
Remove unused test directory

MadLittleMods committed:
Fix focus issue with initial load on mac

Mac apps do not have focus when they first launch. Windows apps do have focus on launch.

MadLittleMods committed:
Update docs on what dependencies are needed to build

MadLittleMods committed:
Update notifier to disable native notification sounds (re: Windows 10)

mikaelbr commented on nwapp/utils/notifier.js:
```diff
@@ -32,10 +35,23 @@ module.exports = function (spec) {
+    // And we set to false because Windows 10 by default makes a separate sound
+    sound: false,
+    // wait with callback until user action is taken on notification
+    wait: false
```
As you are trying to listen for `click` activation on notification, you'll have to wait for the response from the notification. Without waiting you can't get info on if the user clicks or not. So this should be `true`.

MadLittleMods replied:
@mikaelbr Thanks for the info. It seems to work just fine with it being `wait: false`? I believed that option was to make the notification sit there until some user action as taken but that doesn't seem to be the case either.

MadLittleMods replied:
@mikaelbr see https://github.com/mikaelbr/node-notifier/issues/97

MadLittleMods committed:
Add avatars to notifications on windows

MadLittleMods committed:
Start refactor windows installer build to node script

MadLittleMods committed:
Generalize windows desktop build `windows/build.js`

MadLittleMods committed:
Remove extraneous logging from windows build script

MadLittleMods committed:
Remove startup option from windows installer

The option did not obey your choice not to add it to startup which isn't a good experience if you didn't want it enabled. Solves https://github.com/gitterHQ/desktop/issues/106

MadLittleMods committed:
Add context menu option to toggle start on launch

trevorah commented on nwapp/index.js:
```diff
@@ -256,7 +281,13 @@ function showLoggedInWindow(exec) {
-    mainWindowFocused = true;
+    // When a mac app starts up, it doesn't have focus
+    // When a Windows app starts up, it has focus
+    // TODO: Check behaviour on linux
```
can you check it on linux please?

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +36,85 @@ function playNotificationSound() {
-  });
+  var avatarImageUrl = getFullUrlNotificationImageForUri(spec.link);
+
+  getFileAtUrl(avatarImageUrl)
```
Does `node-notifier` not support `https://`?

MadLittleMods replied:
@trevorah Both `nw.js` and `node-notifier` do not support `http(s)` for Windows toast notifications. Needs to be local `file://` urls

trevorah replied:
And they wont crash if they are given the fallback `https://` url?

MadLittleMods replied:
Nope, just a broken image (shows the default Windows no-icon-provided icon image).

We could sniff the environment and only fallback to `http(s)` URL for OSX but this is simpler to understand.

trevorah replied:
Yeah, I agree. The less `if (osx)` code, the better.

trevorah commented on package.json:
```diff
@@ -1,30 +1,32 @@
     "semver": "^4.2.0"
   },
+  "devDependencies": {
+    "bluebird": "^3.2.2",
```
Looks like you're using this in the main app, not just in the build scripts...

MadLittleMods replied:
@trevorah It is included as a `dependency` in the main app: https://github.com/gitterHQ/desktop/pull/100/files#diff-e3600309bfd1f7dbe245b37ce9cd68c2R26

trevorah replied:
Whoops! My bad!

trevorah commented on windows/build.js:
```diff
@@ -0,0 +1,84 @@
+      '"C:\\Program Files (x86)\\Inno Setup 5\\ISCC.exe" "' + path.resolve(__dirname, 'gitter.iss') + '"',
+      '"' + signToolPath + '" sign /f "' + path.resolve(__dirname, '..\\certificates\\troupe-cert.pfx') + '" /p ' + certPassword + ' "' + path.resolve(__dirname, '..\\artefacts\\GitterSetup*') + '"',
+      'rename "' + path.resolve(__dirname, '..\\artefacts\\GitterSetup.exe') + '" "GitterSetup-' + appVersion + '.exe"',
+    ];
```
When constructing command line args, new lines and indents make reading/editing a lot easier.
e.g

``` javascript
var commandList = [
  '"' + signToolPath + '"' +
    ' sign' +
    ' /f "' + path.resolve(__dirname, '..\\certificates\\troupe-cert.pfx') + '"' +
    ' /p ' + certPassword +
    ' "' + path.resolve(__dirname, '..\\opt\\Gitter\\win32\\Gitter.exe') + '"',
  '"C:\\Program Files (x86)\\Inno Setup 5\\ISCC.exe"' +
    ' "' + path.resolve(__dirname, 'gitter.iss') + '"',
  '"' + signToolPath + '"' +
    ' sign' +
    ' /f "' + path.resolve(__dirname, '..\\certificates\\troupe-cert.pfx') + '"' +
    ' /p ' + certPassword + 
    ' "' + path.resolve(__dirname, '..\\artefacts\\GitterSetup*') + '"',
  'rename' +
    '"' + path.resolve(__dirname, '..\\artefacts\\GitterSetup.exe') + '"' +
    ' "GitterSetup-' + appVersion + '.exe"'
];
```

Also, if you are writing a script, _always_ use the full name for flags/arguments in any of your commands. I know that its not possible with these commands, but its just good practice.

trevorah commented on windows/build.bat:
```diff
@@ -1,8 +1,11 @@
+  "C:\Program Files (x86)\Inno Setup 5\ISCC.exe" "%cd%\windows\gitter.iss"
+  "C:\Program Files (x86)\Microsoft SDKs\Windows\v7.1A\Bin\signtool.exe" sign /f "%cd%\certificates\troupe-cert.pfx" /p "%2" "%cd%\artefacts\GitterSetup*"
+  rename "%cd%\artefacts\GitterSetup.exe" "GitterSetup-%1%.exe"
 )
```
So is this file still being used?

MadLittleMods replied:
It is not being used and it can be removed, but I would prefer to keep until the next cycle in this sprint because I almost couldn't find a workaround for https://github.com/nodejs/node/issues/5160 and needed to use it.

It is being replaced by `./windows/build.js`

trevorah commented on windows/build.js:
```diff
@@ -0,0 +1,84 @@
+
+    var commandList = [
+      '"' + signToolPath + '" sign /f "' + path.resolve(__dirname, '..\\certificates\\troupe-cert.pfx') + '" /p ' + certPassword + ' "' + path.resolve(__dirname, '..\\opt\\Gitter\\win32\\Gitter.exe') + '"',
+      '"C:\\Program Files (x86)\\Inno Setup 5\\ISCC.exe" "' + path.resolve(__dirname, 'gitter.iss') + '"',
```
Does Inno Setup always install to the same directory? If not, then the `statResults` are pointless.

MadLittleMods replied:
@trevorah I think you misunderstood this part of the code? We look for the `signtool.exe` on your PC in a variety of locations(which it can be). When any of those checks succeed, `Promise.any(statResults)`, we use the path found to run the build commands.

trevorah replied:
yeah, but if Inno Setup isnt in the right dir, then none of this will work.

MadLittleMods replied:
That is the default location. I do see that you had it as "C:\Program Files\Inno Setup 5\ISCC.exe" but I assume that wasn't on  64-bit system or something? I can add a check for either if you feel it necessary.

trevorah replied:
eh, that'll be fine for now.

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +36,85 @@ function playNotificationSound() {
+        // wait with callback until user action is taken on notification
+        wait: false
+      }, function (err, response) {
+        // response is response from notification
```
If we dont care about the `response` arg, then dont include it.

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +36,85 @@ function playNotificationSound() {
+        // response is response from notification
+        if(err) {
+          throw err;
+        }
```
`if (err) throw err;`

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +36,85 @@ function playNotificationSound() {
+  });
+}
+
 module.exports = function (spec) {
```
I know this isnt you, but please never have a "spec" object with required properties. If an argument is is required, then it should be explicitly an argument. Optional arguments are last or in an options object, thats fine. I'm sure you know this already, but hey.

So, this function is used in three places, in three different ways:

``` javascript
notifier({
  title:   msg.title,
  message: msg.text,
  roomId:  msg.troupeId,
  link:    msg.link,
  click: function () {
    navigateWindowTo(msg.link);
  }
});
```

``` javascript
notifier({
  title: 'Gitter ' + version + ' Available',
  message: 'Head over to gitter.im/apps to update.'
});
```

``` javascript
notifier({
  title: 'Gitter ' + version + ' Available',
  message: 'Click to restart and apply update.',
  click: function() {
    ... (does stuff)
  }
});
```

There are some required arguments and some optional ones (and some ignored ones!). Can you please change the function arguments to reflect this? Also, you may need to check through your existing code in here, as there are some parts that assume that an optional parameter will always be there. This is why the "spec" pattern sucks.

Also, our user_notification "msg" already includes an `icon` property that is a url. That may make your life a little easier...

MadLittleMods replied:
@trevorah Those properties are all optional. I have cleaned them up and made sure we aren't using anything when it isn't available.

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +36,85 @@ function playNotificationSound() {
+    }
+    else {
+      setTimeout(function() {
+        reject(new Error('Took to long to request file', url));
```
"Too"

trevorah replied:
Also, https://github.com/request/request#timeouts.

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +36,85 @@ function playNotificationSound() {
+    // We are not worried if they couldn't fetch the image
+    // as we fallback to the URL itself
+    .catch(function() {
+      return undefined;
```
no need to return, just log something useful.

MadLittleMods replied:
I like being explicit here as the `undefined` is important for a check later(I know it will `return undefined` by default).

trevorah replied:
How about returning `avatarImageUrl`? You would have to change `getFileAtUrl()` to return a url though.

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +36,85 @@ function playNotificationSound() {
+
+// Cache lookup
+var urlFilePathMap = new Map();
+function getFileAtUrl(url) {
```
If you return a file url (`file:///my/image/whatever.png`), then your code that uses this function can be a little bit simpler as it will only have to deal with urls.

trevorah commented on nwapp/index.js:
```diff
@@ -139,6 +155,15 @@ function initApp() {
   });
 
+  events.on('settings:change:launchOnStartup', function(newValue) {
+    console.log('settings:change:launchOnStartup', newValue);
```
`log.info`

MadLittleMods replied:
Sorry, this shouldn't even be in here(was just for testing) but we can log it.

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +36,85 @@ function playNotificationSound() {
+        reject(new Error('Took to long to request file', url));
+      }, 750);
+
+      var tempFileStream = temp.createWriteStream();
```
Fun fact: windows doesn't clean up its temp directory. OS X and Linux does, but only on restart.

I recommend storing them in a specific directory in tmp that you can easily clean up after yourself.

MadLittleMods replied:
@trevorah The `temp.track()` call keeps track of the files and cleans up on exit: https://github.com/gitterHQ/desktop/pull/100/files#diff-d74e5f0a212623b40b7c84b3d6313d78R6

I tested and works fine on Windows.

trevorah replied:
:-1: it's global. We use the temp library in other places.

trevorah replied:
As in it will mess with the autoupdate process.

trevorah commented on windows/build.js:
```diff
@@ -0,0 +1,84 @@
+    var commandRunner = function(command) {
+      return new Promise(function(resolve, reject) {
+        var child = exec(command, function(err, stdout, stderr) {
+          //console.log('stdout', stdout, 'e', err, 'stderr', stderr);
```
delete me

mfonville commented:
With the changed start-up options in this pull; would it also be possible to also add an option to start the app minimized during the autostart?

MadLittleMods commented:
@mfonville Can you please create a separate issue on the repo for this?

mfonville commented:
@MadLittleMods of course, no problem: https://github.com/gitterHQ/desktop/issues/107

MadLittleMods committed:
Add toggle to show in menu bar option for OSX

MadLittleMods committed:
Update v3 PR based on peer review

MadLittleMods commented:
@trevorah Ready for another review pass :grinning:

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -8,11 +22,9 @@ var win = gui.Window.get();
-function getNotificationImageForUri(uri) {
-  if (!uri) return "";
-  if (CLIENT.match(/^osx|^win/)) return "https://avatars.githubusercontent.com/" + uri.split("/")[1] + "?s=32";
+function getFullUrlNotificationImageForUri(uri) {
```
Use the `icon` property from the original faye message. It's a little more future-proof.

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +40,112 @@ function playNotificationSound() {
   });
+events.on('app:quit', function() {
+  log.info('App is quiting, cleanup avatars');
+  rimraf(tempAvatarDirectory)
```
This doesnt block the thread, so the app could quit before this has finished and leave the directory in a weird state. Better make it synchronous as the deletion wont be a massive block (and put it in a try/catch).

MadLittleMods replied:
I thought about this as well and just wasn't sure as it seemed to perform fine in basic testing and thought the node process would wait until any calls finished. I will update just to be safe.

trevorah replied:
Thanks!

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +40,112 @@ function playNotificationSound() {
+      }
+      return cachedFilePath;
+    })
+    .catch(function(cachedFilePath) {
```
Using try/catch for a cache miss is overkill as you're going to be missing the cache pretty often. Put the icon downloader in its own function that returns a promise so you can use it like this:

``` js
function getFileUrlForHttpUrl(url) {
  return ensureCacheReady.then(function() {
    return cache.get(url) || downloadAndCache(url);
  })
  .then(function(filePath) {
    return 'file://' + filePath;
  });
}
```

(I've also simplified some of the function names)

MadLittleMods replied:
Thanks for the simplification! Much better flow

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +40,112 @@ function playNotificationSound() {
+    // as we fallback to the URL itself
+    .catch(function(err) {
+      log.error('Trouble getting avatar for notification' + err.message + err.stack);
+      return undefined;
```
`return avatarImageUrl` as you are in the catch block handling the error. `catch` is the best place to supply fallbacks.

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +40,112 @@ function playNotificationSound() {
+  click: undefined
+};
+
+module.exports = function (options) {
```
`title` and `message` are supplied by every caller of this function. Please change it to `function(title, message, options)` so you dont need to supply defaults for situations that dont even exist.

MadLittleMods replied:
I disagree with this as they are optional. I could easily see a notification with just a title or message.

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +40,112 @@ function playNotificationSound() {
+
+  var avatarImageUrl = getFullUrlNotificationImageForUri(opts.link);
+
+  getFileAtUrl(avatarImageUrl)
```
What happens if `avatarImageUrl` is `undefined` or `''`? Should we even be attempting to run this function if we know that it is going to fail?

MadLittleMods replied:
@trevorah I defaulted it to the `gitterHQ` user avatar.

If someone passes in an explicit dirty value, just let it blow up so they know what to fix and we aren't swallowing their mistakes.

MadLittleMods committed:
Update v3 PR based on feedback, clean up flow

MadLittleMods commented:
@trevorah All fixed up again, ready for another review pass

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +36,109 @@ function playNotificationSound() {
+      }, function (err) {
+        if(err) throw err;
+      });
+      notifier.on('click', function (notifierObject, options) {
```
`notifierObject, options` aren't used. Delete them.

MadLittleMods committed:
Remove unused parameters from notifier click callback

trevorah commented on nwapp/utils/notifier.js:
```diff
@@ -28,14 +36,109 @@ function playNotificationSound() {
+  title: '',
+  message: '',
+  // gitterHQ logo
+  icon: 'https://avatars0.githubusercontent.com/u/5990364?v=3&s=60',
```
Aren't there some operating systems where they show the app icon along with the notification icon? Wont it look a bit weird having two of the same icon? Especially if one icon is slightly different from the other...

MadLittleMods replied:
From my testing, this doesn't seem to happen. It looks like it might for Mac if we made a custom build of `node-notifier` and added our app icon.

OSX:
![](https://i.imgur.com/ZOSgdpA.png)
Windows 10:
![](http://i.imgur.com/HUBKgKN.png)
Ubuntu:
![](https://i.imgur.com/tFsLVnk.png)

trevorah replied:
Damn that icon looks ugly on osx. We can improve that once we actually start supporting osx properly.

Also, use `avatars.githubusercontent.com` and drop the version parameter.

MadLittleMods committed:
Update default notification gitterHQ avatar URL

MadLittleMods committed:
Update notification icon/avatar to use raw absolute path for Windows 8.1

We use the raw absolute path vs `file:///xxx` protocol so notification icons work on Windows 8.1, see https://github.com/mikaelbr/node-notifier/issues/99

MadLittleMods commented:
@trevorah Ready to go. I fixed up the notifications on Windows 8.1 by using raw absolute path vs `file:///xxx` protocol URL, see https://github.com/mikaelbr/node-notifier/issues/99

Just need to go through tesing auto-update with you.

MadLittleMods committed:
Update  task to fix split platform setup introduced earlier

MadLittleMods committed:
Update auto-update params and logic

MadLittleMods committed:
Add `check-path-safety-for-windows` gulp task

Also fixed previous problem when I added `yargs` by installing `normalize-package-data` as a root-level dependency

MadLittleMods committed:
Update auto-update

MadLittleMods committed:
Update readme to document new cli parameters

MadLittleMods commented:
@trevorah I updated the auto-update part of the code and have tested the flow from `2.4.0` -> `3.0.0` -> `3.0.1` on OSX. Can you please take a look before I do a Windows release?

I need some sort of code signing identity in my keychain in order to build the OSX app fully but have been just commenting out that part of the code in the mean time.

MadLittleMods committed:
Add explicit npm 3 depedency and requirement

MadLittleMods committed:
Add comment to `--update-url` section of code

trevorah commented on nwapp/index.js:
```diff
@@ -31,26 +40,62 @@ process.on('uncaughtException', function (err) {
+
+var checkFileExistSync = function(target) {
+  try {
+    return fs.statSync(target);
```
`return !!fs.statSync(target);` so that we return a boolean

trevorah commented on nwapp/index.js:
```diff
@@ -31,26 +40,62 @@ process.on('uncaughtException', function (err) {
+    return false;
+  }
+
+  return false;
```
this will never be reached

MadLittleMods replied:
I'll remove the `return` inside the `catch`

trevorah commented on nwapp/index.js:
```diff
@@ -31,26 +40,62 @@ process.on('uncaughtException', function (err) {
 
-  if (gui.App.argv.length) {
+  var currentInstallPath = argv['current-install-path'];
+  var newUpdaterExecutablePath = argv['new-executable'];
```
Gah, autoupdate is hard. Im not the biggest fan of hitting the disk synchronously on startup if we can avoid it.

``` javascript
var legacyCurrentInstallPath;
var legacyNewUpdaterExecutablePath;
if (argv.length === 2 && checkFileExistSync(argv._[0]) && checkFileExistSync(argv._[1])) {
  // we have a pre v3 desktop app attempting to update to v3+
  // remove after August 2016
  legacyCurrentInstallPath = argv._[0];
  legacyNewUpdaterExecutablePath = argv._[1];
}

var currentInstallPath = argv['current-install-path'] || legacyCurrentInstallPath;
var newUpdaterExecutablePath = argv['new-executable'] || legacyNewUpdaterExecutablePath;
```

trevorah commented on gulpfile.js:
```diff
@@ -225,19 +261,36 @@ gulp.task('autoupdate:push:win', function (done) {
+};
+var platformFolders = [
+  // legacy
+  'desktop',
```
do we want to be pushing to the catch all manifest directory?

MadLittleMods replied:
We can get rid of it, I'm not sure what referenced it before.

trevorah replied:
this is the old path that the app used to ping to check if an update was available.

trevorah commented on gulpfile.js:
```diff
@@ -225,19 +261,36 @@ gulp.task('autoupdate:push:win', function (done) {
+  gulp.task('manifest:push:' + platformFolder, function() {
+    return pushManifestToDest(path.join(platformFolder, 'package.json'));
+  });
+});
```
please dont programatically create build tasks (or public function names in general) as they are a nightmare to find when things go wrong. Its ok to have a little bit of repetition for the sake of clarity.

MadLittleMods replied:
I did this based on existing patterns where we iterate to assemble, ex: `cert:fetch:xxx`, `clean:xxx`, `pack:templates:xxx`, `pack:deb:xxx`, `artefacts:push:xxx`, etc, etc

Do we want to refactor the others? Otherwise leaving it as-is seems to conform to the rest of the file.

trevorah replied:
Writing code in the same style as the rest of the file (or codebase) is really good practice, so I can understand where you are coming from. However, this build file is really poorly written, so im cool with you breaking the implied rules and being explicit with your task names. If anything, this will help with an ongoing refactor of the build.

trevorah commented on gulpfile.js:
```diff
@@ -176,10 +210,12 @@ var dmg_cmd = template('./osx/create-dmg/create-dmg --icon "<%= name %>" 311 50
 
 // Only runs on OSX (requires XCode properly configured)
 gulp.task('sign:osx', ['build'], shell.task([
+  /* */
```
?

MadLittleMods replied:
`gulp build:osx` will fail if you don't have the code signing credentials in your keychain (I don't) so this acts as a nice comment toggle. Would be happy to remove if we get some proper documentation on signing this code and setup on this.

trevorah commented on gulpfile.js:
```diff
@@ -85,13 +88,44 @@ function fetchS3(params, done) {
+
+    // Avoid the high-water mark: https://github.com/gulpjs/gulp/issues/1356
+    return exhaustively(stream);
+});
```
So this can be removed now?

MadLittleMods replied:
In regards to something along the lines of "this can be removed because we use npm 3 that will flatten dependencies"

I'd rather leave the check as it is still possible that it could happen. When you get this failure, it isn't clear what is causing it so the pre-check just gives some peace of mind and catches a potential problem.

MadLittleMods committed:
Updates from v3 review/feedback

MadLittleMods committed:
Fix error when `--update-url` was not defined

MadLittleMods committed:
Fix legacy auto-update logic from mistake in review

MadLittleMods committed:
Update notification handling and auto-update

MadLittleMods committed:
Update auto-update polling logic

Now only pings every 24 hours. Covers https://github.com/gitterHQ/desktop/issues/112

MadLittleMods commented:
@trevorah Polling logic review please, [latest commit](https://github.com/gitterHQ/desktop/commit/5e88e5dc9f22910324d81f88b902a67d4c1fc370)

MadLittleMods committed:
Refactor auto-update logic to not persist

Covers https://github.com/gitterHQ/desktop/issues/112

trevorah commented on nwapp/utils/auto-update.js:
```diff
@@ -1,31 +1,98 @@
   win: 'https://update.gitter.im/win/package.json',
   osx: 'https://update.gitter.im/osx/package.json',
-  linux: 'https://update.gitter.im/linux/package.json'
+  linux: 'https://update.gitter.im/linux64/package.json'
```
err, what about linux32? whats going on here?

MadLittleMods replied:
We just look at the 64bit manifest for whether to update. The other pieces of the linux build are coupled together so this shouldn't be an issue.

But mainly spawned out of the `./nwapp/utils/client-type.js` not distinguishing between the two.

trevorah replied:
thats super confusing. Has this already gone out?

MadLittleMods replied:
Possibly to a few people.

MadLittleMods replied:
I updated this to use `linux32` and `linux64`

trevorah commented on nwapp/utils/auto-update.js:
```diff
@@ -1,31 +1,98 @@
+    if(timeSinceNotify >= notifyUpdateIntervalTime) {
+      cb();
+    }
+  }, notifyUpdatePollingRate);
```
:+1:

trevorah commented on nwapp/utils/auto-update.js:
```diff
@@ -1,31 +1,98 @@
+
+  var newUrl = parsedTargetUrl.toString();
+  return newUrl;
+};
```
normally you would just use environment variables for this.

MadLittleMods replied:
I much prefer command line parameters. Easier to manage cross platform and less environment magic.

trevorah replied:
Thats fair. I guess im used to server side apps.

MadLittleMods committed:
Update auto-update script to split linux32 linux64 manifest checks

Also add `osx` package manifest entry (legacy `mac` entry still in place)

MadLittleMods commented:
Here was my release process. Merge into master (always do releases from master).

Test each platform release as you go. Also check https://gitter.im/apps to make sure the redirects are in place

### OSX
- `gulp build:osx`
- `gulp autoupdate:zip:osx`
- `gulp redirect:source`
- `gulp artefacts:push:osx`
- Don't run `gulp redirect:push:osx` because we are still using the old safari desktop app
- `gulp autoupdate:push:osx`
- `gulp manifest:push:osx`

### Windows:
- `gulp build:linux`
- `gulp cert:fetch:win`
- On a Windows machine: `node ./windows/build.js -p thepfxcertpasswordhere` (I setup network [file sharing from OSX machine](https://support.apple.com/kb/PH18707?locale=en_US) ([make sure the share name doesn't have a space in it](https://github.com/nodejs/node/issues/5160)). Use `pushd`/`popd` for the UNC path `pushd \\ERIC-MACBOOK\eric\Documents\github\desktop`)
- `gulp autoupdate:zip:win`
- `gulp redirect:source`
- `gulp artefacts:push:win`
- `gulp redirect:push:win`
- `gulp autoupdate:push:win`
- `gulp manifest:push:win`

### Linux
- `gulp build:linux`
- `gulp autoupdate:zip:linux`
- `gulp redirect:source`
- `gulp artefacts:push:linux32`
- `gulp artefacts:push:linux64`
- `gulp redirect:push:linux32`
- `gulp redirect:push:linux64`
- `gulp autoupdate:push:linux`
- `gulp manifest:push:linux`
